linalg/blockoperator.hpp

```diff
@@ -172,6 +172,96 @@ class BlockDiagonalPreconditioner : public Solver
    mutable BlockVector yblock;
 };
 
-}
+//! @class BlockLowerTriangularPreconditioner
+/**
+ * \brief A class to handle Block lower triangular preconditioners in a matrix-free implementation.
+ *
+ * Usage:
+ * - Use the constructors to define the block structure
+ * - Use SetBlock to fill the BlockOperator
+ * - Use the method Mult and MultTranspose to apply the operator to a vector.
+ *
+ * If a diagonal block is not set, it is assumed it is an identity block, if an off-diagonal block is not set, it is assumed to be a zero block.
+ *
+ */
+class BlockLowerTriangularPreconditioner : public Solver
+{
+public:
+   //! Constructor for BlockLowerTriangularPreconditioners with the same block-structure for rows and
+   //! columns.
+   /**
+    *  offsets: offsets that mark the start of each row/column block (size
+    *  nRowBlocks+1).  Note: BlockLowerTriangularPreconditioner will not own/copy the data contained
+    *  in offsets.
+    */
+   BlockLowerTriangularPreconditioner(const Array<int> & offsets);
+   //! Constructor for general BlockLowerTriangularPreconditioners.
+   /**
+    *  row_offsets: offsets that mark the start of each row block (size
+    *  nRowBlocks+1).  col_offsets: offsets that mark the start of each column
+    *  block (size nColBlocks+1).  Note: BlockLowerTriangularPreconditioner will not own/copy the
+    *  data contained in offsets.
+    */
+   BlockLowerTriangularPreconditioner(const Array<int> & row_offsets,
+                                      const Array<int> & col_offsets);
+
+   //! Add block op in the block-entry (iblock, iblock).
+   /**
+    * iblock: The block will be inserted in location (iblock, iblock).
+    * op: the Operator to be inserted.
+    */
+   void SetDiagonalBlock(int iblock, Operator *op);
```

Comment:
I don't disagree, although I would note this isn't the case in the BlockDiagonalPreconditioner class.   Would it be more correct to note that the `mult` method of the diagonal blocks should _approximate_ the inverse of the diagonal blocks?  (I could imagine a case where this came from something like SPAI, for example, so it would be a direct multiplication by an approximate inverse